[PM-17562] Add Azure Service Bus for Distributed Events (#540)
bitwarden/contributing-docs: docs/getting-started/server/events.md

## Distributed events (optional)

Events can be distributed via an AMQP messaging system. This messaging system enables new
integrations to subscribe to the events. The system supports either RabbitMQ or Azure Service Bus.

### Listener / Handler pattern

The goal of moving to distributed events is to build additional service integrations that consume
events. To make it easy to support multiple AMQP services (RabbitMQ and Azure Service Bus), the act
of listening to the stream of events is decoupled from the act of responding to an event.

**Listeners**

> - One ListenerService per communication platform (i.e. one for RabbitMQ, one for Azure Service Bus)

Reviewer: ```suggestion
- One listener per communication platform (e.g. one for RabbitMQ, one for Azure Service Bus).
```

> - Multiple instances can be configured to run independently, each with its own Handler and

Reviewer: ```suggestion
- Multiple instances can be configured to run independently, each with its own handler and
```

>   subscription/queue.

Reviewer: ```suggestion
  subscription / queue.
```

> - Perform all the aspects of setup/teardown, subscription, etc. for the messaging platform, but do

Reviewer: ```suggestion
- Perform all the aspects of setup / teardown, subscription, etc. for the messaging platform, but do
```

  not directly process any events themselves. Instead, they delegate to the handler with which they
  are configured.

**Handlers**

> - One EventHandler per integration (e.g. currently HttpPost, EventRepository)

Reviewer: ```suggestion
- One handler per integration (e.g. HTTP post or event database repository).
```

- Completely isolated from and know nothing of the messaging platform in use. This allows them to be
  freely reused across different communication platforms.
- Perform all aspects of handling an event.
- Allows them to be highly testable as they are isolated and decoupled from the more complicated
  aspects of messaging.

This combination allows for a configuration inside of `Startup.cs` that pairs instances of the

> ListenerService for the currently running messaging platform with any number of Handlers. It also

Reviewer: ```suggestion
listener service for the currently running messaging platform with any number of handlers. It also
```

> allows for quick development of new Handlers as they are focused only on the task of handling a

Reviewer: ```suggestion
allows for quick development of new handlers as they are focused only on the task of handling a
```

        "password": "SET_A_PASSWORD_HERE_123",
        "exchangeName": "events-exchange",
        "eventRepositoryQueueName": "events-write-queue",
        "httpPostQueueName": "events-httpPost-queue",
      }
      "httpPostUrl": "<HTTP POST URL>",
    }
    ```

2.  (optional) The `httpPostQueueName` and `httpPostUrl` specified above are optional. If they are
    defined, an `HttpPostEventHandler` will be added to a `RabbitMqEventListenerService` instance
    that will `POST` the event to the configured URL.

>     - Tip: [RequestBin](http://requestbin.com/) provides an easy to set up server that will receive

Reviewer: 🎨 Missed this but there should be some kind of info box we can use here instead of "Tip:".

With these changes in place, you should see the database events written as before, but you'll also
see in the RabbitMQ management interface that the messages are flowing through the configured
exchange/queues.

### Azure Service Bus implementation

> The Azure Service Bus implementation is a configurable replacement for Azure Queue. Instead of

Reviewer: 💭 Wondering about the "replacement" language here but I also don't want to document a migration per se.

Author: Would "configurable alternative" work better? So it's more just another option rather than a wholesale replacement?

writing Events to the queue to be picked up, they are sent to the configured service bus topic. An
instance of `AzureServiceBusEventListenerService` is then configured with the
`AzureTableStorageEventHandler` to subscribe to that topic and write Events to Azure Table Storage.

> Similar to RabbitMQ above, the end result is the same (events are stored in Azure Table Storage),

Reviewer: ❓ Or the database right? Not just Azure Table Storage.

Author: Right now, the Azure Service Bus implementation assumes it will use an `AzureTableStorageEventHandler` which is only setup to write to Azure Table Storage. If we want, we could add the ability to configure it for DB storage, or someone can change `startup.cs` where it configures which handler to use.

Bit of background/context:
The `EventRepositoryHandler` gets an injected `IEventRepository`, so it theoretically could just write to any EventRepository that happens to be injected. But the TableStorage `EventRepository` assumes `IEvent` to be implemented by `EventTableEntity` (i,e. not `EventMessage`), which is why `AzureTableStorageEventHandler` calls `EventTableEntity.IndexEvent(eventMessage)` before passing it off to the `EventRepository`. 

I think it's probably reasonable to refactor that TableStorage side to make that call instead, rather than throwing an exception if it's not `EventTableEntity`. That would allow for us to use the original `EventRepositoryHandler` and do all of the magic via injection. But I also wanted to stay on the side of safety with this PR and it seemed like a bad idea to modify that Table Storage piece.

but the addition of the service bus topic allows for other integrations to subscribe.

As with the RabbitMQ implementation above, an `HttpPostEventHandler` can be configured to run and
POST events to a URL via a separate subscription.

```kroki type=mermaid
graph TD
	  subgraph With Service Bus
        B1[EventService]
        B2[AzureServiceBusEventWriteService]
        B3[Azure Service Bus Topic]
        B4[AzureTableStorageEventHandler]
        B5[HttpPostEventHandler]
        B6[Events in Azure Tables]
        B7[HTTP Server]

        B1 -->|IEventWriteService| B2 --> B3
        B3-->|AzureServiceBusEventListenerService| B4 --> B6
        B3-->|AzureServiceBusEventListenerService| B5
        B5 -->|HTTP POST| B7
    end

    subgraph With Storage Queue
        A1[EventService]
        A2[AzureQueueHostedService]
        A3[Azure Storage Queue]
        A4[AzureQueueHostedService]
        A5[Events in Azure Tables]

        A1 -->|IEventWriteService| A2 --> A3 -->|RepositoryEventWriteService| A4 --> A5

end
```

#### Running the Azure Service Bus emulator

> 1. Make sure you have Azurite setup locally (as

Reviewer: ```suggestion
1. Make sure you have Azurite set up locally (as
```

   [per the normal instructions](https://contributing.bitwarden.com/getting-started/server/guide#azurite)

>    for writing Events to Azure Table Storage). In addition, this assumes you're using the `mssql`

Reviewer: ```suggestion
   for writing events to Azure Table Storage). In addition, this assumes you're using the `mssql`
```

   default profile and have the `${MSSQL_PASSWORD}` set via `.env`.

> 2. Run docker compose to add/start the local emulator:

Reviewer: ```suggestion
2. Run Docker Compose to add/start the local emulator:
```

Not sure if this is needed in other places too.

```bash
docker compose --profile servicebus up -d
```

> - Note: The service bus emulator waits 15 seconds before starting. You can check the console in

Reviewer: ```suggestion
- Note: The Service Bus emulator waits 15 seconds before starting. You can check the console in
```

🎨 Use an info box too.

>   Docker desktop to make sure it's up before launching the server.

Reviewer: 🎨 Consider describing this for cases when a GUI isn't available.

   for writing events to Azure Table Storage). In addition, this assumes you're using the `mssql`
   default profile and have the `${MSSQL_PASSWORD}` set via `.env`.

2. Run Docker Compose to add/start the local emulator:

```bash
docker compose --profile servicebus up -d
```

:::info

The service bus emulator waits 15 seconds before starting. You can check the console in Docker
desktop or run `docker logs service-bus` to verify the service is up before launching the server.

:::

#### Configuring the server to use Azure Service Bus for events

1. Add the following to your `secrets.json` in `dev` to configure the service bus:

```json
	"eventLogging": {
	  "azureServiceBus": {
		"connectionString": "\"Endpoint=sb://localhost;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;\"",
		"topicName": "event-logging",
		"eventRepositorySubscriptionName": "events-write-subscription",
	  }
	},
```

2. Re-run the secrets script to publish the new secrets

```bash
pwsh setup_secrets.ps1 -clear
```

3. Start or re-start all services, including `EventsProcessor`.

> ### Configuring the Webhook Handler (optional)

Reviewer: ```suggestion
### Configuring the webhook (optional)
```
